posix: use sys_sem instead of k_spinlock for pool synch

cfriedt opened the pull request:
Based on Andy's talk at eoss 2024, use the `sys/sem.h` api instead of the `spinlock.h` api to synchronize pooled elements since it has minimal overhead like the k_sem api but also works (better) from userspace

fgrandel commented on include/zephyr/sys/sem.h:
```diff
@@ -138,6 +138,66 @@ int sys_sem_take(struct sys_sem *sem, k_timeout_t timeout);
  */
 unsigned int sys_sem_count_get(struct sys_sem *sem);
 
+/**
```
You might want to introduce this helper in a separate commit, so it gets more visibility, can be explained in the commit message and be attributed to the right subsys?

cfriedt replied:
Done!

fgrandel commented on include/zephyr/sys/sem.h:
```diff
@@ -138,6 +138,66 @@ int sys_sem_take(struct sys_sem *sem, k_timeout_t timeout);
+ * exactly once:
+ *
+ * @code{.c}
+ * for (sys_sem_take(&sem); ...; sys_sem_take(&sem)) {
```
The third param is probably meant to be sys_sem_give(...).

cfriedt replied:
Fixed!

fgrandel commented on lib/posix/options/key.c:
```diff
@@ -127,19 +128,21 @@ int pthread_key_create(pthread_key_t *key,
-	k_spinlock_key_t key_key;
 
-	key_key = k_spin_lock(&pthread_key_lock);
+	ret = sys_sem_take(&pthread_key_lock, K_FOREVER);
```
It's probably obvious and I just can't see it: Why don't you use your new contruct here and in the other cases below?

cfriedt replied:
Done - it was mainly just travelling, being busy, and laziness. I didn't quite get a chance to fully look at all of the cases until this morning :-)

Commit:
sys_sem: add a for-loop-like macro similar to spinlocks

Add SYS_SEM_LOCK() and SYS_SEM_LOCK_BREAK which are mostly the
same as K_SPINLOCK() and K_SPINLOCK_BREAK.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
clang-format: add entry for SYS_SEM_LOCK()

SYS_SEM_LOCK() is a for-loop-like macro similar inspired by
K_SPINLOCK().

Add an entry to .clang-format so that autoformatting works as
expected.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
posix: use sys_sem instead of k_spinlock for pool synch

Based on Andy's talk at eoss 2024, use the sys/sem.h api instead
of the spinlock.h api to synchronize pooled elements since it
has minimal overhead like semaphores but also works from
userspace.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

andyross commented on include/zephyr/sys/sem.h:
```diff
@@ -138,6 +139,82 @@ int sys_sem_take(struct sys_sem *sem, k_timeout_t timeout);
+	__ASSERT(*rc == 1, "SYS_SEM_LOCK exited with goto, break or return, "
+			   "use SYS_SEM_LOCK_BREAK instead.");
+}
+#define SYS_SEM_LOCK_ONEXIT __attribute__((__cleanup__(z_sys_sem_lock_onexit)))
```
Seems like if we're going to repeat this paradigm we should think about up-levelling it so that an abstracted "lock" can provide the preprocessor tools to define these loop/block macros with  ~one-line defintions or whatever.  But that's for a different PR.

andyross reviewed (approved):
Looks clean.  Though to be fair I ended up skipping the slide with the discussion on this.  Basically POSIX spinlocks are semi-dangerous in ported code.  Userspace spinlocking in Linux (where the processes are preemptable!) is deep voodoo done by only Serious High Performance Computing Applications (stuff like C10k database servers, etc...), who are prepared to handle lock breaking and backoff themselves.  POSIX can't do it alone.

But... we don't really have that problem, and sys_sem avoids the biggest gotchas.  So if we're going to provide this, at least we can say ours is less of a footgun than linux/glibc's.

cfriedt commented:
This doesn't switch the POSIX spinlock over to sys_sem yet - here we just "avoid spinlocks in library code" in the synchronization of pools where there is a possibility that it could be called from userspace.

I'll make a separate PR to convert pthread spinlocks to use sys_sem under the hood too.
